aws/s3_access_point.go

```diff
@@ -0,0 +1,418 @@
+	wg := new(sync.WaitGroup)
+	wg.Add(len(identifiers))
+	errChans := make([]chan error, len(identifiers))
+	for i, s3Mrap := range identifiers {
+		errChans[i] = make(chan error, 1)
+		go deleteS3MultiRegionAccessPointAsync(wg, errChans[i], svc, s3Mrap, accountId)
+	}
+	wg.Wait()
+
+	// Collect all the errors from the async delete calls into a single error struct.
+	var allErrs *multierror.Error
+	for _, errChan := range errChans {
+		if err := <-errChan; err != nil {
+			allErrs = multierror.Append(allErrs, err)
+			logging.Logger.Debugf("[Failed] %s", err)
+			telemetry.TrackEvent(commonTelemetry.EventContext{
+				EventName: "Error Nuking S3 Multi Region Access Point",
+			}, map[string]interface{}{})
+		}
+	}
+	finalErr := allErrs.ErrorOrNil()
+	if finalErr != nil {
+		return errors.WithStackTrace(finalErr)
+	}
+
+	// Now wait until the S3 multi region access point are deleted
+	err = retry.DoWithRetry(
+		logging.Logger,
+		"Waiting for all S3 multi region access points to be deleted.",
+		// Wait a maximum of 5 minutes: 10 seconds in between, up to 30 times
+		30, 10*time.Second,
+		func() error {
+			areDeleted, err := areAllS3MultiRegionAccessPointsDeleted(svc, identifiers, accountId)
+			if err != nil {
+				return errors.WithStackTrace(retry.FatalError{Underlying: err})
+			}
+			if areDeleted {
+				return nil
+			}
+			return fmt.Errorf("Not all S3 multi region access points deleted.")
+		},
+	)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	for _, s3Mrap := range identifiers {
+		logging.Logger.Debugf("[OK] S3 Multi Region Access Point %s was deleted", aws.StringValue(s3Mrap))
+	}
+	return nil
+}
+
+// areAllS3MultiRegionAccessPointsDeleted returns true
+func areAllS3MultiRegionAccessPointsDeleted(svc *s3control.S3Control, identifiers []*string, accountId string) (bool, error) {
+	for _, identifier := range identifiers {
+		resp, err := svc.GetMultiRegionAccessPoint(&s3control.GetMultiRegionAccessPointInput{
+			AccountId: aws.String(accountId),
+			Name:      identifier,
+		})
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NoSuchMultiRegionAccessPoint" { // s3 multi region access point deleted
+				continue
+			} else {
+				return false, err
+			}
+		}
+
+		if resp != nil {
+			return false, nil
+		}
+	}
+
+	// At this point, all the S3 multi region access points are either nil, or deleted.
+	return true, nil
+}
+
+// deleteS3MultiRegionAccessPointAsync deletes the provided S3 Multi Region Access Point asynchronously ina goroutine,
+// using wait groups for concurrency control and a return channel for errors.
+func deleteS3MultiRegionAccessPointAsync(wg *sync.WaitGroup, errChan chan error, svc *s3control.S3Control, s3Mrap *string, accountId string) {
+	defer wg.Done()
+
+	input := &s3control.DeleteMultiRegionAccessPointInput{
+		AccountId: aws.String(accountId),
+		Details: &s3control.DeleteMultiRegionAccessPointInput_{
```

Comment:
Is there an extra `_` here at the end of `PointInput`? :thinking: `Details: &s3control.DeleteMultiRegionAccessPointInput_{`

Reply:
Yes. aws-sdk-go (v1) docs says `A container for the information associated with a DeleteMultiRegionAccessPoint request`.

[https://docs.aws.amazon.com/sdk-for-go/api/service/s3control/#DeleteMultiRegionAccessPointInput_](https://docs.aws.amazon.com/sdk-for-go/api/service/s3control/#DeleteMultiRegionAccessPointInput_)

Reply:
Ah great, TIL - thank you!